# Multiple Regression 

In [2]:
#Multiple regression is like linear regression, but with more than one independent value, meaning that we try to predict a value based on two or more variables.

In [5]:
#Use cars.csv
#Contains info about cars
#we will import the csv using the pandas module

In [7]:
import pandas as pd
df = pd.read_csv("../datasets/cars.csv")
df.head()

Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105

In [8]:
# Then make a list of the independent values and call this variable X.

# Put the dependent values in a variable called y.

In [9]:
X = df[['Weight', 'Volume']]
y = df['CO2']

# Tip: It is common to name the list of independent values with a upper case X, and the list of dependent values with a lower case y.

In [12]:
# From the sklearn module we will use the LinearRegression() method to create a linear regression object.

# This object has a method called fit() that takes the independent and dependent values as parameters and fills the regression object with data that describes the relationship:

from sklearn import linear_model

In [13]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [14]:
#Now we have a regression object that are ready to predict CO2 values based on a car's weight and volume:

In [15]:
#predict the CO2 emission of a car where the weight is 2300kg, and the volume is 1300cm3:
predictedCO2 = regr.predict([[2300, 1300]])

In [16]:
predictedCO2

array([107.2087328])

In [17]:
#We have predicted that a car with 1.3 liter engine, and a weight of 2300 kg, will release approximately 107 grams of CO2 for every kilometer it drives.

## Checking the correctness
## This can be done by determining the Coefficient.
## The Multiple Regression Coefficient is a Factor that describes the relationship with an unknown variable 
## For Example 
Example: if x is a variable, then 2x is x two times. x is the unknown variable, and the number 2 is the coefficient.

In this case, we can ask for the coefficient value of weight against CO2, and for volume against CO2. The answer(s) we get tells us what would happen if we increase, or decrease, one of the independent values.

In [21]:
import pandas as pd
from sklearn import linear_model

df = pd.read_csv("../datasets/cars.csv")

X = df[['Weight', 'Volume']]
y = df['CO2']

regr = linear_model.LinearRegression()
regr.fit(X, y)

print(regr.coef_)

[0.00755095 0.00780526]


The result array represents the coefficient values of weight and volume.

Weight: 0.00755095
Volume: 0.00780526

These values tell us that if the weight increase by 1kg, the CO2 emission increases by 0.00755095g.

And if the engine size (Volume) increases by 1 cm3, the CO2 emission increases by 0.00780526 g.

## Test the prediction 
I think that is a fair guess, but let test it!

We have already predicted that if a car with a 1300cm3 engine weighs 2300kg, the CO2 emission will be approximately 107g.

What if we increase the weight with 1000kg?

In [22]:
import pandas
from sklearn import linear_model

df = pandas.read_csv("../datasets/cars.csv")

X = df[['Weight', 'Volume']]
y = df['CO2']

regr = linear_model.LinearRegression()
regr.fit(X, y)

predictedCO2 = regr.predict([[3300, 1300]])

print(predictedCO2)

[114.75968007]


We have predicted that a car with 1.3 liter engine, and a weight of 3300 kg, will release approximately 115 grams of CO2 for every kilometer it drives.

Which shows that the coefficient of 0.00755095 is correct:

107.2087328 + (1000 * 0.00755095) = 114.75968